gh-113666: Adding missing UF_ and SF_ flags to module 'stat'

ronaldoussoren opened the pull request:
Add some constants to module 'stat' that are used on macOS.

<!-- gh-issue-number: gh-113666 -->
* Issue: gh-113666
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--113667.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

ronaldoussoren committed:
gh-113666: Adding missing UF_ and SF_ flags to module 'stat'

Add some constants to module 'stat' that are used on macOS.

serhiy-storchaka commented on Lib/stat.py:
```diff
@@ -111,18 +112,31 @@ def S_ISWHT(mode):
+UF_COMPRESSED = 0x00000020 # macOS: file is compressed
+UF_TRACKED   = 0x00000040  # macOS: used for handling document IDs
+UF_DATAVAULT = 0x00008000  # macOS: entitlement needed for I/O
+UF_HIDDEN    = 0x00008000  # macOS: file should not be displayed
```
Why are they equal?

ronaldoussoren replied:
Probably because I'm bad at copy&paste :-(

serhiy-storchaka commented on Lib/stat.py:
```diff
@@ -111,18 +112,31 @@ def S_ISWHT(mode):
+    # Group of bits in st_flags, specific to Apple platforms.
+
+    UF_SETTABLE = 0xffff      # owner settable flags
+    SF_SUPPORTED = 0x9f0000   # superuser supported flags
```
```suggestion
    UF_SETTABLE  = 0x0000ffff # owner settable flags
    SF_SUPPORTED = 0x009f0000 # superuser supported flags
```

serhiy-storchaka commented on Lib/stat.py:
```diff
@@ -111,18 +112,31 @@ def S_ISWHT(mode):
+
+    UF_SETTABLE = 0xffff      # owner settable flags
+    SF_SUPPORTED = 0x9f0000   # superuser supported flags
+    SF_SETTABLE  = 0x3ffff000 # superuser settable flags
```
It is suspicious that `UF_SETTABLE & SF_SETTABLE != 0`.

ronaldoussoren replied:
I intend to add a test for macOS that tests against the actual platform values for every constant in `stat`, even if that might not survive review.

ronaldoussoren replied:
> It is suspicious that `UF_SETTABLE & SF_SETTABLE != 0`.

This was a copy&paste error as well.

ronaldoussoren committed:
Update Lib/stat.py

Co-authored-by: Serhiy Storchaka <storchaka@gmail.com>

ronaldoussoren committed:
Add test that tests values of constants on macOS

ronaldoussoren committed:
Fix definitions

ronaldoussoren committed:
Add new definitions to _stat.c as well

ronaldoussoren committed:
Merge branch 'gh-113666' of github.com:ronaldoussoren/cpython into gh-113666

ronaldoussoren committed:
Fix typo

ronaldoussoren committed:
Add new flags to the generic tests as well.

ronaldoussoren committed:
Fix whitespace at end of line

serhiy-storchaka commented on Doc/library/stat.rst:
```diff
@@ -374,10 +382,46 @@ The following flags can be used in the *flags* argument of :func:`os.chflags`:
+
+   .. versionadded: 3.13
+
+.. data:: UF_SETTABLE
```
```suggestion
.. data:: SF_SETTABLE
```

ronaldoussoren committed:
Merge branch 'main' into gh-113666

ronaldoussoren committed:
BSD and older macOS

1. SF_SETTABLE and UF_SETTABLE are available on BSD as well

   Add generic definitions to stat.py and fallback defines to
   _stat.c (the latter matching how we handle other generic flags)

2. Older versions of macOS do not have SF_SUPPORTED and SF_SYNTHETIC

   _stat.c defines those for all macOS platforms, and redefines
   SF_SETTABLE to match recent macOS versions.

   I did this primarily to end up with an explainable situation: our
   installers are build on recent versions of macOS and can be used
   on older versions.

   The implementation in this changeset ensures we end up with the same
   definitions regardless of how the binary was build (old SDK or new SDK
   and old deployment target)

serhiy-storchaka commented on Lib/test/test_stat.py:
```diff
@@ -241,6 +241,18 @@ def test_module_attributes(self):
+
+        for flag in self.file_flags:
+            if flag.startswith("UF"):
+                self.assertTrue(getattr(self.statmod, flag) & self.statmod.UF_SETTABLE, f"{flag} notin UF_SETTABLE")
```
```suggestion
                self.assertTrue(getattr(self.statmod, flag) & self.statmod.UF_SETTABLE, f"{flag} not in UF_SETTABLE")
```

serhiy-storchaka commented on Modules/_stat.c:
```diff
@@ -579,6 +604,7 @@ stat_exec(PyObject *module)
     ADD_INT_MACRO(module, UF_TRACKED);
     ADD_INT_MACRO(module, UF_DATAVAULT);
     ADD_INT_MACRO(module, UF_HIDDEN);
+    ADD_INT_MACRO(module, SF_SETTABLE);
```
#ifdef SF_SETTABLE

serhiy-storchaka commented on Modules/_stat.c:
```diff
@@ -226,10 +230,22 @@ typedef unsigned short mode_t;
 
+#ifndef SF_SETTABLE
+#  define SF_ARCHIVED 0xffff0000
+#endif
```
Why is this needed?

ronaldoussoren replied:
Sigh, I've check the commit a couple of times and still missed that I forgot to update the name here.

ronaldoussoren committed:
Fix typo

ronaldoussoren committed:
Update Lib/test/test_stat.py

Co-authored-by: Serhiy Storchaka <storchaka@gmail.com>

ronaldoussoren committed:
fix ci errors

ronaldoussoren committed:
Merge branch 'gh-113666' of github.com:ronaldoussoren/cpython into gh-113666

ronaldoussoren committed:
Correct documentatation

serhiy-storchaka reviewed (approved):
LGTM.

I only wondering why define default values for these flags both in `_stat.c` and `stat.py`? Usually, the C implementation only expose constants if they are defined:
```c
#ifdef SOMETHING
    ADD_INT_MACRO(module, SOMETHING);
#endif
```
Then the Python implementation can define the default value and override it with import from the C module if it is defined, or keep it undefined.
```py
SOMETHING = ...
try:
    from _somemodule import SOMETHING
except ImportError:
    pass
```
or
```py
try:
    from _somemodule import SOMETHING
except ImportError:
    SOMETHING = ...
```
Defining it in the C code if it is undefined is pretty unusual.
